src/puppetlabs/puppetdb/cli/shovel.clj

```diff
@@ -0,0 +1,378 @@
+             (.setSchedulerSupport true)
+             (.setPersistent true)
+             (enable-jmx true)
+             (set-usage! memory-usage #(.getMemoryUsage %) "MemoryUsage")
+             (set-usage! store-usage #(.getStoreUsage %) "StoreUsage")
+             (set-usage! temp-usage #(.getTempUsage %) "TempUsage"))
+        mc (doto (.getManagementContext mq)
+             (.setCreateConnector false))
+        db (doto (.getPersistenceAdapter mq)
+             (.setIgnoreMissingJournalfiles true)
+             (.setArchiveCorruptedIndex true)
+             (.setCheckForCorruptJournalFiles true)
+             (.setChecksumJournalFiles true))]
+    mq))
+
+(defn ^BrokerService start-broker!
+  "Starts up the supplied broker, making it ready to accept
+  connections."
+  [^BrokerService broker]
+  (doto broker
+    (.start)
+    (.waitUntilStarted)))
+
+(defn stop-broker!
+  "Stops the supplied broker"
+  [^BrokerService broker]
+  (doto broker
+    (.stop)
+    (.waitUntilStopped)))
+
+(defn ^BrokerService retry-build-and-start-broker!
+  [brokername dir config]
+  (try
+    (start-broker! (build-embedded-broker brokername dir config))
+    (catch java.io.EOFException e
+      (log/error e (i18n/trs "EOF Exception caught during broker start, this might be due to KahaDB corruption. Consult the PuppetDB troubleshooting guide."))
+      (throw e))))
+
+(defn ^BrokerService build-and-start-broker!
+  "Builds ands starts a broker in one go, attempts restart upon known exceptions"
+  [brokername dir config]
+  (try
+    (try
+      (start-broker! (build-embedded-broker brokername dir config))
+      (catch java.io.EOFException e
+        (log/warn (i18n/trs "Caught EOFException on broker startup, trying again. This is probably due to KahaDB corruption (see \"KahaDB Corruption\" in the PuppetDB manual)."))
+        (retry-build-and-start-broker! brokername dir config)))
+    (catch java.io.IOException e
+      (throw (java.io.IOException.
+              (i18n/trs "Unable to start broker in {0}. This is probably due to KahaDB corruption or version incompatibility after a PuppetDB downgrade (see \"KahaDB Corruption\" in the PuppetDB manual)." (pr-str dir))
+              e)))))
+
+(defn activemq-connection-factory [spec]
+  (proxy [PooledConnectionFactory java.lang.AutoCloseable]
+      [(ActiveMQConnectionFactory. spec)]
+    (close [] (.stop this))))
+
+;; Questionable code
+
+(defn upgrade-lockfile-path [vardir]
+  (str (io/file vardir "upgraded_mq.lock")))
+
+(defn needs-upgrade?
+  "Returns true if the user has an unmigrated ActiveMQ director in
+  their vardir"
+  [config]
+  (let [vardir (get-in config [:global :vardir])]
+    (and (fs/exists? (io/file vardir "mq"))
+         (not (fs/exists? (upgrade-lockfile-path vardir))))))
+
+(defn lock-upgrade
+  "Puts a lockfile in vardir to indicate the user has already migrated
+  to Stockpile"
+  [config]
+  (fs/touch (upgrade-lockfile-path (get-in config [:global :vardir]))))
```

Comment:
(If we happen to end up with paths due to any of the above, note that there's a (.toFile path), since iirc reynes/fs doesn't know how to handle paths.)